pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(resource).ToNot(BeIdenticalTo(expected))
+		})
+
+		It("should return no error on success", func() {
+			_, err := createFederatedResource(scheme, resource)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("the returned resource's Kind should be of Federated<kind>", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			expectedKind := federatedKindPrefix + resource.GroupVersionKind().Kind
+			Expect(expected.GroupVersionKind().Kind).
+				To(Equal(expectedKind))
+		})
+
+		It("the returned resource's Group should be the default KubeFed FederatedGroup", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GroupVersionKind().Group).
+				To(Equal(kubefedopt.DefaultFederatedGroup))
+		})
+
+		It("the returned resource's Version should be the default KubeFed FederatedVersion", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GroupVersionKind().Version).
+				To(Equal(kubefedopt.DefaultFederatedVersion))
+		})
+
+		It("the returned resource's name should be the same as the input's object", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GetName()).
+				To(Equal(resource.GetName()))
+		})
+
+		It("the returned resource's template should be the input's unstructured", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			expected, _ = createFederatedResource(scheme, resource)
+
+			mapObj, found, err := getTemplateField(expected)
+			targetResource := &unstructured.Unstructured{}
+
+			convErr := scheme.Convert(resource, targetResource, nil)
+			Expect(convErr).ToNot(HaveOccurred())
+			removeUnwantedFields(targetResource)
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj["spec"]).To(BeEquivalentTo(targetResource.Object["spec"]))
+			Expect(mapObj["metadata"]).To(BeEquivalentTo(targetResource.Object["metadata"]))
+		})
+
+		It("the returned resource's placement should be an empty matcher", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			expected, _ = createFederatedResource(scheme, resource)
+
+			mapObj, found, err := getMatchLabelsField(expected)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj).To(BeEmpty())
+		})
+
+		It("the returned resource's namespace should be the input's namespace", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GetNamespace()).
+				To(Equal(resource.GetNamespace()))
+		})
+	})
+}
+
+func testDistribute() {
+	var (
+		resource      *corev1.Pod
+		fedResource   *unstructured.Unstructured
+		f             *federator
+		initObjs      []runtime.Object
+		clusterNames  []string
+		scheme        *runtime.Scheme
+		clientPatcher client.Client
+	)
```

Comment:
There's more we can do here to simplify and reduce the use/scope of closures but I can look at it later and submit a separate PR. No need to hold this one up for that.

Reply:
Sure, just be careful and take in mind the future-proof argument I gave you before and also keep things clear.